src/plugins/gpgme/gpgme.c

```diff
@@ -0,0 +1,194 @@
+{
+	const Key * metaEncrypt = keyGetMeta (k, ELEKTRA_GPGME_META_ENCRYPT);
+	if (metaEncrypt && strcmp (keyString (metaEncrypt), "1") == 0)
+	{
+		return 1;
+	}
+	return 0;
+}
+
+/**
+ * @brief checks if a given Key k is in the spec namespace.
+ * @retval 0 if the Key k is in the spec namespace.
+ * @retval 1 if the Key k is NOT in the spec namespace.
+ */
+static inline int isSpecNamespace (const Key * k)
+{
+	return (keyGetNamespace (k) == KEY_NS_SPEC);
+}
+
+/**
+ * @brief Encrypt all Keys in the KeySet "data" that have a meta-key "gpg/encrpyt" set.
+ * @see isMarkedForEncryption (const Key * k)
+ * @param handle holds the plugin configuration.
+ * @param data the KeySet to be encrypted.
+ * @param errorKey will hold and error description if an operation fails.
+ * @retval 1 on success.
+ * @retval -1 on failure. In this case errorKey will provide a description.
+ */
+static int gpgEncrypt (ELEKTRA_UNUSED Plugin * handle, KeySet * data, ELEKTRA_UNUSED Key * errorKey)
+{
+	Key * k;
+	// TODO KeySet * pluginConfig = elektraPluginGetConfig (handle);
+
+	// TODO convert pluginConfig to an array of recipient keys
+
+	ksRewind (data);
+	while ((k = ksNext (data)) != 0)
+	{
+		if (!isMarkedForEncryption (k) || isSpecNamespace (k))
+		{
+			continue;
+		}
+
+		// TODO use gpgme to encrypt the value of k
+	}
+	return 1;
+}
+
+/**
+ * @brief Decrypt all Keys in the KeySet "data" that have a meta-key "gpg/encrpyt" set.
+ * @see isMarkedForEncryption (const Key * k)
+ * @param handle holds the plugin configuration.
+ * @param data the KeySet to be decrypted.
+ * @param errorKey will hold and error description if an operation fails.
+ * @retval 1 on success.
+ * @retval -1 on failure. In this case errorKey will provide a description.
+ */
+static int gpgDecrypt (ELEKTRA_UNUSED Plugin * handle, KeySet * data, ELEKTRA_UNUSED Key * errorKey)
+{
+	Key * k;
+	// TODO KeySet * pluginConfig = elektraPluginGetConfig (handle);
+
+	gpgme_ctx_t ctx;
+	gpgme_error_t err;
+
+	err = gpgme_new (&ctx);
+
+	ksRewind (data);
+	while ((k = ksNext (data)) != 0)
+	{
+		if (!isMarkedForEncryption (k) || isSpecNamespace (k))
+		{
+			continue;
+		}
+
+		// TODO use gpgme to decrypt the value of k
+		gpgme_data_t in;
+		gpgme_data_t out;
+		gpgme_decrypt_result_t dec_result;
+
+		// TODO error handling
+		// TODO consider original data type (string, binary, null)
+
+		err = gpgme_data_new_from_mem (&in, keyValue (k), keyGetValueSize (k), 0);
+		err = gpgme_data_new (&out);
+
+		err = gpgme_op_decrypt (ctx, out, in);
+		dec_result = gpgme_op_decrypt_result (ctx);
+		// TODO handle gppme errors
+
+		// TODO finish here
+
+		gpgme_data_release (out);
+		gpgme_data_release (in);
+	}
+	gpgme_release (ctx);
+	return 1;
+}
+
+int elektraGpgmeOpen (ELEKTRA_UNUSED Plugin * handle, ELEKTRA_UNUSED Key * errorKey)
+{
+	gpgme_error_t err;
+
+	gpgme_check_version (NULL);
+	setlocale (LC_ALL, "");
+	gpgme_set_locale (NULL, LC_CTYPE, setlocale (LC_CTYPE, NULL));
```

Comment:
We should not globally change anything (and especially not locales, this might make applications dysfunctional). If we set it, it should only be done for the gpgme context but why set it at all?

It looks like gpgme is engineered properly, as far as I saw it everything seems to be operating only on its context (and never globally, except in gpgme_set_locale if a NULL pointer is passed).

It might be useful to allocate the gpgme context only once (in open), it might be faster then.

Reply:
The `gpgme` manual recommends to set the locale for initializing `gpgme`. I am not sure why they recommend it.

Reply:
It's fine for prototyping but probably should not be used like this in any released software.
What stands against creating a gpgme context to solve the issue (and not have to rely on maybe wonky locale)?

Reply:
> What stands against creating a gpgme context to solve the issue (and not have to rely on maybe wonky locale)?

It would lead to yet another configuration setting of the plugin. If the locales are broken, pinentries shown by other applications will be broken, too, won't they?

But if we set the locales for the context is a minor thing. I only wanted to note that we must not set it globally.